File: python/tictactoe.py
from dataclasses import dataclass
from enum import Enum
from typing import Literal


class Status(Enum):
    x = 1
    o = 0
    DRAW = 2
    IN_PROGRESS = 3


@dataclass(frozen=True)
class State:
    x: int = 0b000000000
    o: int = 0b000000000


Player = Literal[1, 0]


def move(state: State, choice: int, player: Player) -> State:
    choice = 1 << choice
    assert not ((state.x | state.o) & choice)
    match player:
        case 1:
            return State(x=state.x | choice, o=state.o)
        case 0:
            return State(x=state.x, o=state.o | choice)


def check(state):
    horizontal = 0b111000000
    vertical = 0b100100100
    diag1 = 0b100010001
    diag2 = 0b001010100
    draw = 0b111111111
    for i in (0, 3, 6):
        if state.x & (horizontal >> i) == (horizontal >> i):
            return Status.x
        if state.o & (horizontal >> i) == (horizontal >> i):
            return Status.o
    for i in (0, 1, 2):
        if state.x & (vertical >> i) == (vertical >> i):
            return Status.x
        if state.o & (vertical >> i) == (vertical >> i):
            return Status.o
    if (state.x & diag1 == diag1) or (state.x & diag2 == diag2):
        return Status.x
    if (state.o & diag1 == diag1) or (state.o & diag2 == diag2):
        return Status.o
    if draw == state.x | state.o:
        return Status.DRAW
    return Status.IN_PROGRESS


def available_moves(state: State):
    not_occupied = ~(state.x | state.o)
    empty = not_occupied & 0b111111111
    return list(filter(lambda i: (empty & (1 << i)), range(9)))


def print_board(state: State):
    board = [" " for _ in range(9)]
    for i in range(9):
        if state.x & (1 << i):
            board[i] = "x"
        elif state.o & (1 << i):
            board[i] = "o"
    ref = "0|1|2\n3|4|5\n6|7|8\n\n"
    board_string = f"{board[0]}|{board[1]}|{board[2]}\n{board[3]}|{board[4]}|{board[5]}\n{board[6]}|{board[7]}|{board[8]}"
    print(ref + board_string)


"""
b 8 7 6 5 4 3 2 1 0
x 0 0 0 0 0 1 0 0 0
o 0 0 0 0 0 0 0 0 0
"""

File: C/tictactoe.c
#include <stdio.h>
#include "tictactoe.h"

void printBoard(State* state){
    char board[9];
    for(int i = 0; i < 9; i++){
        if(state->playerX & (1<<i)){
            board[i] = 'X';
        }
        else{
            if(state->playerO & (1<<i)){
                board[i] = 'O';
            }
            else{
                board[i] = ' ';
            }
        }
    }
    printf("0|1|2\n3|4|5\n6|7|8\n\n");
    printf("%c|%c|%c\n%c|%c|%c\n%c|%c|%c\n",
        board[0], board[1], board[2],
        board[3], board[4], board[5],
        board[6], board[7], board[8]
    );
}

int makeMove(State *state, int player, int move) {
    if (move < 0 || move > 8) return 0;
    int bit = 1 << move;
    if ((state->playerX | state->playerO) & bit) return 0;  // already occupied
    if (player == 1)
        state->playerX |= bit;
    else
        state->playerO |= bit;
    return 1;
}

int available_moves(State* state){
    return (~(state->playerX | state->playerO)) & 0b111111111;
}

Status check(State* state){
    static const int horizontal = 0b111000000;
    static const int vertical   = 0b100100100;
    static const int diag1      = 0b100010001;
    static const int diag2      = 0b001010100;
    static const int full       = 0b111111111;

    /* Horizontal rows */
    for (int i = 0; i <= 6; i += 3) {
        int mask = horizontal >> i;
        if ((state->playerX & mask) == mask){ 
            return STATUS_X; 
        }
        if ((state->playerO & mask) == mask){ 
            return STATUS_O;
        }
    }
 
    /* Vertical columns */
    for (int i = 0; i <= 2; i++) {
        int mask = vertical >> i;
        if((state->playerX & mask) == mask){ 
            return STATUS_X;
        }
        if((state->playerO & mask) == mask){ 
            return STATUS_O;
        }
    }
 
    /* Diagonals */
    if ((state->playerX & diag1) == diag1 || (state->playerX & diag2) == diag2) {
        return STATUS_X;
    }
    if ((state->playerO & diag1) == diag1 || (state->playerO & diag2) == diag2) {
        return STATUS_O;
    }
 
    /* Draw */
    if (full == (state->playerX | state->playerO)){ 
        return STATUS_DRAW;
    }
 
    return STATUS_IN_PROGRESS;
}


void clearScreen(){
    printf("\033[2J\033[1;1H");
}